composer/utils/data.py

```diff
@@ -95,26 +96,61 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
     return image_tensor, target_tensor
 
 
-def add_dataset_transform(dataset, transform):
+def add_dataset_transform(dataset, transform, location="end", pre_post: str = "pre") -> Dataset:
     """Flexibly add a transform to the dataset's collection of transforms.
 
     Args:
         dataset: A torchvision-like dataset
         transform: Function to be added to the dataset's collection of transforms
+        location (str, torchvision transform), optional: Where to insert the transform in
+            the sequence of transforms. "end" will append to the end or "start" will
+            insert at index 0, a transform (e.g. torchvision.transforms.ToTensor) will
+            insert pre- or post- the first instance of `location`, determined by the
+            `pre_post` argument. Default = "end". If `location` is not in list of
+            transforms, will append to end.
+        pre_post (str), optional: Whether to insert before ("pre") or after ("post") the
+            transform string passed in 'location'. Ignored if transform = "start" or "end". Default = "pre".
 
     Returns:
         The original dataset. The transform is added in-place.
     """
 
-    if not hasattr(dataset, "transform"):
-        raise ValueError(f"Dataset of type {type(dataset)} has no attribute 'transform'. Expected TorchVision dataset.")
+    if not hasattr(dataset, transform):
+        raise AttributeError(textwrap.dedent(f"""Dataset must have attribute `transform`."""))
+
+    valid_non_transform_locations = ["start", "end"]
+    valid_pre_post_values = ["pre", "post"]
+    if location not in valid_non_transform_locations and pre_post not in valid_pre_post_values:
+        raise ValueError(
+            textwrap.dedent(
+                f"Invalid value combination for argument `pre_post`: `{pre_post} and `location`: {location}. If location is not one of ['start', 'end'], `pre_post` must be one of ['pre', 'post']."
+            ))
 
     if dataset.transform is None:
         dataset.transform = transform
-    elif hasattr(dataset.transform, "transforms"):  # transform is a Compose
-        dataset.transform.transforms.append(transform)
+    # Check if dataset.transform is of type transforms.Compose and ensure idempotency by not adding transform if it's already present
+    elif isinstance(dataset.transform,
```

Comment:
This is becoming sufficiently complex that perhaps we want to add some simple testing around this.

Reply:
Also, can this line be broken into multiple lines?